File: authserver/src/main/java/l2s/authserver/network/gamecomm/gs2as/SetAccountInfo.java
package l2s.authserver.network.gamecomm.gs2as;

import io.netty.buffer.ByteBuf;
import l2s.authserver.accounts.SessionManager;
import l2s.authserver.accounts.SessionManager.Session;
import l2s.authserver.network.gamecomm.GameServerDescription.HostInfo;
import l2s.authserver.network.gamecomm.ReceivablePacket;
import l2s.authserver.network.gamecomm.vertx.GameServerConnection;
import org.apache.commons.lang3.ArrayUtils;

public class SetAccountInfo extends ReceivablePacket {
    private final String _account;
    private final int _size;
    private final int[] _deleteChars;

    public SetAccountInfo(GameServerConnection client, ByteBuf byteBuf) {
        super(client, byteBuf);

        _account = readString(byteBuf);
        _size = byteBuf.readUnsignedByte();
        int size = byteBuf.readInt();
        if (size > 7 || size <= 0)
            _deleteChars = ArrayUtils.EMPTY_INT_ARRAY;
        else {
            _deleteChars = new int[size];
            for (int i = 0; i < _deleteChars.length; ++i)
                _deleteChars[i] = byteBuf.readInt();
        }
    }

    @Override
    protected void runImpl(GameServerConnection client) {
        if (client.isAuthed()) {
            Session session = SessionManager.getInstance().getSessionByName(_account);
            if (session == null)
                return;
            for (HostInfo host : client.getGameServerDescription().getHosts())
                session.getAccount().addAccountInfo(host.getId(), _size, _deleteChars);
        }
    }
}

File: authserver/src/main/java/l2s/authserver/network/gamecomm/gs2as/SetAccountInfo.kt
package l2s.authserver.network.gamecomm.gs2as

import io.netty.buffer.ByteBuf
import l2s.authserver.accounts.SessionManager
import l2s.authserver.network.gamecomm.GameServerDescription
import l2s.authserver.network.gamecomm.ReceivablePacket
import l2s.authserver.network.gamecomm.vertx.GameServerConnection
import org.apache.commons.lang3.ArrayUtils

/**
 * @author VISTALL
 * @date 20:52/25.03.2011
 */
class SetAccountInfo(client: GameServerConnection, byteBuf: ByteBuf) : ReceivablePacket(client, byteBuf) {

    private val _account: String? = readString(byteBuf)
    private val _size = byteBuf.readUnsignedByte()
    private val _deleteChars: IntArray

    init {
        val size: Int = byteBuf.readInt()
        _deleteChars = if (size > 7 || size <= 0)
            ArrayUtils.EMPTY_INT_ARRAY
        else {
            IntArray(size) {
                byteBuf.readInt()
            }
        }
    }

    override fun runImpl(client: GameServerConnection) {
        val gs: GameServerDescription = client.description
        if (client.isAuthed) {
            val session =
                SessionManager.getInstance().getSessionByName(_account) ?: return
            for (host in gs.hosts)
                session.account.addAccountInfo(host.id, _size.toInt(), _deleteChars)
        }
    }
}